packages/peregrine/lib/talons/CartPage/GiftOptions/useGiftOptions.js

```diff
@@ -0,0 +1,206 @@
+    const [{ cartId }] = useCartContext();
+
+    const [setGiftOptions] = useMutation(SET_GIFT_OPTIONS_QUERY);
+
+    /**
+     * Start of Temp stuff.
+     *
+     * This is only needed till GQL has gift options coverage.
+     */
+    const restUrl = useMemo(() => {
+        if (isSignedIn) {
+            return '/rest/default/V1/carts/mine/gift-message';
+        } else {
+            return `/rest/default/V1/guest-carts/${cartId}/gift-message`;
+        }
+    }, [cartId, isSignedIn]);
+    const [, { sendRequest }] = useRestApi(restUrl);
+
+    /**
+     * Updating backend using rest because GQL coverage
+     * for gift options is un available at this point.
+     */
+    const updateUsingRest = useCallback(
+        newGiftOptions => {
+            const data = {
+                gift_message: {
+                    message: newGiftOptions.gift_message,
+                    extension_attributes: {
+                        wrappingAddPrintedCard:
+                            newGiftOptions.include_printed_card,
+                        wrappingAllowGiftReceipt:
+                            newGiftOptions.include_gift_receipt
+                    }
+                }
+            };
+
+            return sendRequest({
+                options: {
+                    method: 'POST',
+                    body: JSON.stringify(data)
+                }
+            });
+        },
+        [sendRequest]
+    );
+    /**
+     * End of Temp stuff.
+     */
+
+    const updateGiftOptions = useCallback(
+        optionsToUpdate => {
+            const newGiftOptions = {
+                cart_id: cartId,
+                include_gift_receipt: includeGiftReceipt,
+                include_printed_card: includePrintedCard,
+                gift_message: giftMessage,
+                ...optionsToUpdate
+            };
+            setGiftOptions({
+                variables: newGiftOptions
+            });
+            /**
+             * TODO remove this after GQL gets gift options coverage.
+             */
+            updateUsingRest(newGiftOptions);
+        },
+        [
+            updateUsingRest,
+            setGiftOptions,
+            cartId,
+            includeGiftReceipt,
+            includePrintedCard,
+            giftMessage
+        ]
+    );
+
+    /**
+     * Throttling message update. Only make 1 mutation
+     * every 10 seconds. This is to save on bandwidth.
+     *
+     * More info: https://lodash.com/docs/4.17.15#throttle
+     */
+    const throttledMessageUpdate = useMemo(() => {
+        return throttle(
+            (updateGiftOptions, newGiftMessage) => {
+                updateGiftOptions({
+                    gift_message: newGiftMessage
+                });
+            },
+            10000,
+            {
+                leading: false
+            }
+        );
+    }, []);
+
+    const updateGiftMessage = useCallback(
+        newGiftMessage => {
+            setGiftMessage(newGiftMessage);
```

Comment:
I'm not sure you actually need to use a state value here. Think about this like the apollo cache _is_ the state. Updates to the cache (after you do some mutation) are automatically picked up by the query hook, which causes a render update.

Reply:
As I mentioned in the other comment, waiting for a mutation to change the state of a UI component creates a bad user experience. Mutations on an average are taking around 600-800ms in Venia as of today, that is not an acceptable wait to check a box. My point of using a local state is to create a better UI experience over a couple of state variables.

We can use optimistic updates from apollo as well which I have explored briefly. Our apollo usage is not matured enough to handle optimistic updates. Doing it as part of this PR is not valid. I deserves to be its own task.

Let me know if I am missing something here.

Reply:
You're right -- I was thinking about this from the perspective of how I implemented "coupons". Those have a button that, when clicked, displays a wait state while the mutation resolves.

This is a bit different though since UX didn't provide direction about any submit button, which means this would have to submit on blur, or after some timeout :/